yt/utilities/fortran_utils.py

```diff
@@ -199,10 +189,12 @@ def read_vector(f, d, endian="="):
             vec_size,
         )
     vec_num = int(vec_len / vec_size)
-    if isinstance(f, file):  # Needs to be explicitly a file
+    if isinstance(f, io.TextIOBase):
```

Comment:
This one I'm not totally sure about.

> The TextIOBase ABC extends IOBase. It deals with streams whose bytes represent text, and handles encoding and decoding to and from strings. TextIOWrapper, which extends TextIOBase, is a buffered text interface to a buffered raw stream (BufferedIOBase). Finally, StringIO is an in-memory stream for text.

I think the check here is whether it's a filelike object; according to the numpy docs:

> file or str or Path - Open file object or filename.

(also accepts path objects)

I don't think we actually want to change this to `io.TextIOBase`, since we're explicitly really just looking for whether or not it's something that can be supplied to `np.fromfile`.  `np.frombuffer).  `io.BufferedIOBase` is not the opposite of this -- that is triggered when the file / stream object has buffering enabled, but `frombuffer` is really only interested in if something has the buffer interface.

I think the original `if` logic is okay, even if we change over the `fromstring` to `frombuffer`.  Basically, we're attempting to determine whether we need to call `read` on the object or not; I think the check here should be, is this a file, or is it an IOBase that requires `read`.

Reply:
The problem with the original logic is that is was built on the fact that `file` was a builtin type in Python 2. `file` was documented as "the type of the object returned by `open()`, which is `io.TextIOBase` in Python 3.

what's deprecated here is `np.fromfile`'s ability to read binary buffers

> I don't think we actually want to change this to io.TextIOBase, since we're explicitly really just looking for whether or not it's something that can be supplied to np.fromfile

I've inspected the source code upstream, but it's written in C and it's really not obvious to me how it translates...

> I think the check here should be, is this a file, or is it an IOBase that requires read.

do you mean filename here ? as I said, `file` isn't a thing in Python 3 :/

Reply:
Nope, I mean file.

The question for me isn't whether it's a file or not -- it's whether it's a buffer or something that needs to be `read` first.

The two cases are essentially, "is this an open file handle / IO handle / etc", or is this a blob that's already been read in, that we're parsing through bit-by-bit.

Reply:
OK I must confess I feel really out of my league here. Weirdly I never could get my head around the differences that seem to be important here. Would you be able to make a working suggestion, or to push to my branch ? (if so, please feel free to !)

Reply:
I think, after reading the `io` docs, that what we want is to check if it's a subclass of `io.IOBase`.  If it is, then we call `read` on it first and feed that to frombuffer.  Otherwise, we pass it to `frombuffer` directly, since it's more expensive to check if it actually exposes the buffer interface.  (And skip the else at the end, because it's OK for numpy to be the source of our error.)

The difference is essentially if we're doing:

```
f = open( ... )
read_vec(f)
```

or

```
buf = f.read(...)
read_vec(buf)
```

Reply:
ok I'll patch this then, thanks a lot !

Reply:
Here it goes !